Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
                 if (this.Tip != null)
                     break;
 
+                // TODO: this rewind code may never happen. 
+                // The node will complete loading before connecting to peers so the  
+                // chain will never know if a reorg happened.
                 utxoHash = this.UTXOSet.Rewind().GetAwaiter().GetResult();
             }
             this.Puller.SetLocation(this.Tip);
 
+            this.asyncLoop = this.asyncLoopFactory.Run($"Consensus Loop", async token =>
+            {
+                await this.PullerLoop(this.nodeLifetime.ApplicationStopping);
+            }, 
+            this.nodeLifetime.ApplicationStopping, 
+            repeatEvery: TimeSpans.RunOnce);
+
             this.logger.LogTrace("(-)");
         }
 
-        public IEnumerable<BlockResult> Execute(CancellationToken cancellationToken)
+        /// <summary>
+        /// Dispose components in <see cref="ConsensusLoop"/>.
+        /// </summary>
+        public void Stop()
         {
-            while (true)
+            this.asyncLoop?.Dispose();
+        }
+
+        /// <summary>
+        /// A puller method that will continuously loop and ask for the next block  in the chain from peers.
+        /// The block will then be passed to the consensus validation. 
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that will stop the loop.</param>
+        /// <returns>The task the loop will be running under.</returns>
+        private Task PullerLoop(CancellationToken cancellationToken)
```

Comment:
why is this Task? why not void? there is no awaiting and you just return completed task, it does not make much sense to me

Reply:
the cancellation can be thrown up the async loop would handle it well i think, so i don't see any reason for task here

Reply:
I removed it (this will just come back when we make this async).

Reply:
yes, i think it is better to have incremental progress and not to have half-baked things that "wait for later use". when we need it, the change is trivial, but should we forget about it or should we later come with different evolution of the code, we would leave a technical debt and we should avoid that